### PR TITLE
Microfon/Camera off indication

### PR DESCRIPTION
**Pull request by daniel-jodlos**
image

https://membraneframework.atlassian.net/browse/MV-42

### CODE REVIEW

File: assets/src/room.ts [hunk 1]
```diff
@@ -66,17 +71,21 @@ export class Room {
...
-            );
+          let audio_ids = this.localAudioStream?.getTracks().map((track) =>
+            this.webrtc.addTrack(track, this.localAudioStream!, {
+              active: true,
+            })
+          );
+
+          if (audio_ids != undefined) {
+            this.localAudioTrackIds = this.localAudioTrackIds.concat(audio_ids);
+          }
```
Comment (bblaszkow06) [inline assets/src/room.ts:82]: This is quite incocsistent with tthe way `localVideoTrackId`
```suggestion
          let audio_ids = this.localAudioStream?.getTracks().forEach((track) =>
            const trackId = this.webrtc.addTrack(track, this.localAudioStream!, {
              active: true,
            });
            this.localAudioTrackIds.push(trackId);
          );
```

File: assets/src/room.ts [hunk 2]
```diff
@@ -105,6 +114,15 @@ export class Room {
             });
           }
           this.tracks.get(ctx.peer.id)?.push(ctx);
+
+          if (ctx.track?.kind === "audio") {
+            setVisibilityOfMuteIndicator(ctx.peer.id, ctx.metadata.active);
+          } else if (
+            ctx.track?.kind === "video" &&
+            !(ctx.metadata && ctx.metadata.type === "screensharing")
```
Comment (bblaszkow06) [inline assets/src/room.ts:122]:
```suggestion
            ctx.track?.kind === "video" && ctx.metadata?.type !== "screensharing"
```
You may also change the condition in line 104.

File: assets/src/room.ts [hunk 3]
```diff
@@ -66,17 +71,22 @@ export class Room {
...
-            ?.getTracks()
-            .forEach((track) =>
-              this.webrtc.addTrack(track, this.localAudioStream!, {})
+          this.localAudioStream?.getTracks().forEach((track) => {
+            const trackId = this.webrtc.addTrack(
+              track,
+              this.localAudioStream!,
+              {
+                active: true,
+              }
```
Comment (bblaszkow06) [inline assets/src/room.ts:80]: Fix formatting pls
```suggestion
              { active: true }
```

**Review by bblaszkow06 — DISMISSED** [review]

**Review by bblaszkow06 — APPROVED** [review]

File: assets/src/room_ui.ts [hunk 4]
```diff
@@ -26,6 +26,53 @@ type MediaStreams = {
...
+  } else {
+    image?.classList.remove("invisible");
+  }
+}
+
+export function setNoVideoMsgVisible(peer_id: string, active: boolean) {
+  let text = document.querySelector(
+    `#feed-${peer_id} div[name='no-video-msg']`
+  );
+  console.log(text);
```
Comment (mickel8) [inline assets/src/room_ui.ts:59]: Let's delete it

File: assets/src/room_ui.ts [hunk 5]
```diff
@@ -26,6 +26,53 @@ type MediaStreams = {
...
+  audioButton.addEventListener("click", () =>
+    callback(!getAudioButtonStatus())
+  );
+}
+
+export function addVideoStatusChangedCallback(
+  callback: (status: boolean) => any
+) {
+  videoButton.addEventListener("click", () =>
+    callback(!getVideoButtonCurrentStatus())
```
Comment (mickel8) [inline assets/src/room_ui.ts:41]: Why do we negate current status? Also, let's call this `getVideoButtonStatus` to be consistent with `getAudioButtonStatus`.

File: lib/videoroom_web/templates/room/index.html.heex [hunk 6]
```diff
@@ -22,13 +22,17 @@
                     <audio></audio>
                     <video class="w-full h-full"></video>
                     <div name="video-label" class="absolute text-white text-shadow-lg bottom-0 left-0 p-2">Placeholder</div>
+                    <img class="invisible invert" src={static_path(@conn, "/svg/mic-off-fill.svg")} style="position: absolute; left: 0; top: 0; width: 10%, height: 10%">
+                    <div name="no-video-msg" class="invisible absolute text-white w-full h-full top-0 left-0" style="text-align: center">No video feed</div>
```
Comment (mickel8) [inline lib/videoroom_web/templates/room/index.html.heex:26]: Could we make order of `<div>  <div>` the same as for the `remote-video-feed-template` or this was done on purpose?

File: assets/src/room_ui.ts [hunk 7]
```diff
@@ -26,6 +26,53 @@ type MediaStreams = {
...
+  );
+}
+
+export function setMuteIndicatorVisible(peer_id: string, active: boolean) {
```
Comment (mickel8) [inline assets/src/room_ui.ts:45]: I would consider treating microphone and camera in the same way and call these functions

`setMicIndicator` and `setCamIndicator`.

Current naming `setMuteIndicatorVisible` suggests that this function always set indicator to be visible which is not true.

**Review by mickel8 — CHANGES_REQUESTED** [review]

File: assets/src/room_ui.ts (hunk 5 above)
Comment (daniel-jodlos) [reply assets/src/room_ui.ts:41]: This function add a callback for a status change. The button doesn't automatically change the state, you need a callback to do it. The negation is effectively status change logic

Comment (daniel-jodlos) [reply assets/src/room_ui.ts:41]: This function add a callback for a status change. The button doesn't automatically change the state, you need a callback to do it. The negation is effectively status change logic

**Review by mickel8 — APPROVED** [review]
One more thing, could we change `No video feed` to something like `Camera turned off` or `Camera muted`?

**Pull request merged**